---
title: "Identifying marker genes in LM-MP Macrophages"
author: "Stephanie Hicks"
output: 
    html_document:
        toc: true
        toc_float: true
        highlight: tango
        number_sections: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval=FALSE)
```

# Preprocessing and Normalization

Load packages
```{r loadpackages, message=FALSE, warning=FALSE, eval=TRUE}
dataPath <- "/users/shicks1/data/kulkarni"
workingPath <- "/users/shicks1/projects/kulkarni"
```

## Create `SingleCellExperiment` object


Load data and create a `SingleCellExperiment` object. 
Remove genes not expressed in any cell. Extract `mgi_symbol` 
gene name`. 
```{r sceobject, eval=FALSE}
library(monocle)
library(scater)
cds <- readRDS(file.path(dataPath, "Kulkarni_cds.rds")) # CellDataSet class in monocle

# Convert to a SingleCellExperiment
gene_df <- DataFrame(Gene = rownames(exprs(cds)))
rownames(gene_df) <- gene_df$Gene
sce <- SingleCellExperiment(assays = list(counts = exprs(cds)), 
                            colData = pData(cds), rowData = gene_df)

# Remove genes not expressed in any cell
keep_feature <- rowSums(counts(sce) > 0) > 0
table(keep_feature)
sce <- sce[keep_feature, ]

# calcualte QC metrics
sce <- calculateQCMetrics(sce)
head(colData(sce))

# extract mgi gene symbol name
sce <- scater::getBMFeatureAnnos(sce,
    filters = "ensembl_gene_id",
    attributes = c("ensembl_gene_id",
        "mgi_symbol", "chromosome_name",
        "start_position", "end_position"),
    feature_symbol = "mgi_symbol",
    feature_id = "ensembl_gene_id",
    biomart = "ENSEMBL_MART_ENSEMBL",
    dataset = "mmusculus_gene_ensembl",
    host = "www.ensembl.org")

# define feature names in `feature_symbol` column
rownames(sce) <- rowData(sce)$feature_symbol
saveRDS(sce, file = file.path(dataPath,"Kulkarni_sce.rds"))
```

## Create a Seurat object 
Because we want to try using the Seurat pipeline, 
we will convert to a Seurat object. 
```{r}
library(Seurat)
sce <- readRDS(file.path(dataPath,"Kulkarni_sce.rds")) 
sce <- sce[!duplicated(row.names(sce)), ]

# Set up treatment (TH) object
trtTH <- CreateSeuratObject(raw.data = SummarizedExperiment::assay(sce[,
                            SummarizedExperiment::colData(sce)$treatment == "TH"], "counts"), 
                            project = "TRT_TH", min.cells = 5)
trtTH@meta.data <- data.frame(trtTH@meta.data, 
                              as.data.frame(SummarizedExperiment::colData(sce[,
                                        SummarizedExperiment::colData(sce)$treatment == "TH"])))
trtTH <- FilterCells(trtTH, subset.names = "nGene", low.thresholds = 500, high.thresholds = Inf)
trtTH <- NormalizeData(trtTH)
trtTH <- ScaleData(trtTH, display.progress = F)

# Set up treatment (TL) object
trtTL <- CreateSeuratObject(raw.data = SummarizedExperiment::assay(sce[,
                                        SummarizedExperiment::colData(sce)$treatment == "TL"], "counts"), 
                            project = "TRT_TL", min.cells = 5)
trtTL@meta.data <- data.frame(trtTL@meta.data, 
                              as.data.frame(SummarizedExperiment::colData(sce[,
                                        SummarizedExperiment::colData(sce)$treatment == "TL"])))
trtTL <- FilterCells(trtTL, subset.names = "nGene", low.thresholds = 500, high.thresholds = Inf)
trtTL <- NormalizeData(trtTL)
trtTL <- ScaleData(trtTL, display.progress = F)
```

### Use CCA to correct for batch effect 

Gene selection for input to CCA
```{r}
trtTH <- FindVariableGenes(trtTH, do.plot = F)
trtTL <- FindVariableGenes(trtTL, do.plot = F)
g.1 <- head(rownames(trtTH@hvg.info), 1000)
g.2 <- head(rownames(trtTL@hvg.info), 1000)
genes.use <- unique(c(g.1, g.2))
genes.use <- intersect(genes.use, rownames(trtTH@scale.data))
genes.use <- intersect(genes.use, rownames(trtTL@scale.data))
```

Perform CCA. Identifies common sources of variation between two data sets. 
the `RunCCA()` function will also combine two objects into a single object 
and store the canonical correlation vectors (the vectors that project each 
dataset into the maximally correlated subspaces). We also store the 
original dataset identity as a column in `object@meta.data`. 
```{r}
trt.combined <- RunCCA(trtTH, trtTL, genes.use = genes.use, num.cc = 30)

trt.combined@meta.data$treatment <- 
  factor(trt.combined@meta.data$treatment, 
         levels=c("TH", "TL"), labels=c("EnzymeTH", "EnzymeTL"))
```

Align the CCA subspaces
```{r}
trt.combined <- AlignSubspace(trt.combined, reduction.type = "cca", grouping.var = "treatment", 
    dims.align = 1:25)
```

Confirm batch correction worked with t-SNE
```{r}
trt.combined <- RunTSNE(trt.combined, reduction.use = "cca.aligned", dims.use = 1:25, 
    do.fast = T)
TSNEPlot(trt.combined, do.return = T, pt.size = 0.5, group.by = "treatment")
TSNEPlot(trt.combined, do.return = T, pt.size = 0.5, group.by = "cell_type")
```


# Finding marker genes 

## Unsupervised clustering to identify macrophages

Let's do a quick clustering to identify macrophge population 
using a small resolution of 0.25
```{r}
trt.combined <- FindClusters(trt.combined, reduction.type = "cca.aligned", 
    resolution = 0.25, dims.use = 1:25, force.recalc=TRUE)
TSNEPlot(trt.combined, do.label = T, do.return = T, pt.size = 0.5)
```

Looks like the macrophages are labeled as the `0` 
population. Next, filter for only macrophages and 
create a separate Seurat object
```{r}
trt.combined.macro <- SubsetData(trt.combined, 
           cells.use = trt.combined@cell.names[trt.combined@meta.data$res.0.25 == 0])
```

Visualize T-SNE
```{r}
trt.combined.macro <- RunTSNE(trt.combined.macro, reduction.use = "cca.aligned", dims.use = 1:25, 
    do.fast = T)
TSNEPlot(trt.combined.macro, do.return = T, pt.size = 0.5, group.by = "treatment")
TSNEPlot(trt.combined.macro, do.return = T, pt.size = 0.5, group.by = "cell_type")
TSNEPlot(trt.combined.macro, do.label = T, do.return = T, pt.size = 0.5)
```


### Sub-clustering (resolution of 0.25)

Ok, good. Now's let's do some clustering at a 
small resolution again 0.25 (but only on a 
subset of the cells) because we are interested 
only in the macrophage population. 
```{r}
trt.combined.macro <- FindClusters(trt.combined.macro, reduction.type = "cca.aligned", 
    resolution = 0.25, dims.use = 1:25, force.recalc=TRUE)
TSNEPlot(trt.combined.macro, do.return = T, pt.size = 0.5, group.by = "treatment")
TSNEPlot(trt.combined.macro, do.label = T, do.return = T, pt.size = 0.5)
```

Looks like there are two strong clusters. 
We could explore other resolutions too. 
```{r}
trt.combined.macro <- FindClusters(trt.combined.macro, reduction.type = "cca.aligned", 
    resolution = 0.40, dims.use = 1:25, force.recalc=TRUE)
TSNEPlot(trt.combined.macro, do.return = T, pt.size = 0.5, group.by = "treatment")
TSNEPlot(trt.combined.macro, do.label = T, do.return = T, pt.size = 0.5)
```

This one looks like there are 4 clusters. 
For now let's go with the two. 

#### Differential expression (resolution of 0.25)

Next, we will identify marker genes for 
cluster 0 and 1 using the resolution of 0.25

```{r}
trt.combined.macro <- FindClusters(trt.combined.macro, reduction.type = "cca.aligned", 
    resolution = 0.25, dims.use = 1:25, force.recalc=TRUE)
```

#### Cluster 0

```{r}
clust0.markers <- FindConservedMarkers(trt.combined.macro, ident.1 = 0,
                                       grouping.var = "treatment", 
                                       print.bar = FALSE)
head(clust0.markers)

## remove genes with negative relationship. Those will be
##     identified cluster 1
dat <- clust0.markers[clust0.markers$EnzymeTL_avg_logFC > 0 & 
                      clust0.markers$EnzymeTH_avg_logFC > 0 & 
                      clust0.markers$EnzymeTL_p_val_adj < 0.05 & 
                      clust0.markers$EnzymeTH_p_val_adj < 0.05, ]
dat <- dat[order(rowMeans(dat[, c("EnzymeTL_avg_logFC", "EnzymeTH_avg_logFC")]), 
      decreasing = TRUE),]
dim(dat) # 75 DE genes 
head(dat, 20)

write.csv(dat, file.path(workingPath, 
                         "markergenes/res25/seurat_macrophagesonly_clust0markers.csv"), 
          quote = FALSE, row.names=TRUE)

pdf(file.path(workingPath, 
              "figs/res25/seurat_macrophagesonly_clust0markers.pdf"), 
    width=12, height=12)
FeaturePlot(object = trt.combined.macro, features.plot = row.names(dat)[1:20], 
            min.cutoff = "q9", cols.use = c("lightgrey", "blue"), pt.size = 0.5)
dev.off()

```


#### Cluster 1

```{r}
clust1.markers <- FindConservedMarkers(trt.combined.macro, ident.1 = 1,
                                       grouping.var = "treatment", 
                                       print.bar = FALSE)
head(clust1.markers)

## remove genes with negative relationship. Those will be
##     identified cluster 0
dat <- clust1.markers[clust1.markers$EnzymeTL_avg_logFC > 0 & 
                      clust1.markers$EnzymeTH_avg_logFC > 0 & 
                      clust1.markers$EnzymeTL_p_val_adj < 0.05 & 
                      clust1.markers$EnzymeTH_p_val_adj < 0.05,]
dat <- dat[order(rowMeans(dat[, c("EnzymeTL_avg_logFC", "EnzymeTH_avg_logFC")]), 
      decreasing = TRUE),]
dim(dat) # 148 DE genes 
head(dat, 20)

write.csv(dat, file.path(workingPath, 
                         "markergenes/res25/seurat_macrophagesonly_clust1markers.csv"), 
          quote = FALSE, row.names=TRUE)

pdf(file.path(workingPath, 
              "figs/res25/seurat_macrophagesonly_clust1markers.pdf"), 
    width=12, height=12)
FeaturePlot(object = trt.combined.macro, features.plot = row.names(dat)[1:20], 
            min.cutoff = "q9", cols.use = c("lightgrey", "blue"), pt.size = 0.5)
dev.off()

```



### Sub-clustering (resolution of 0.40)

Above we considered another resolution of 
0.40, which looked like there might be four 
clusters. Let's explore that.

```{r}
trt.combined.macro <- FindClusters(trt.combined.macro, reduction.type = "cca.aligned", 
    resolution = 0.40, dims.use = 1:25, force.recalc=TRUE)
TSNEPlot(trt.combined.macro, do.return = T, pt.size = 0.5, group.by = "treatment")
TSNEPlot(trt.combined.macro, do.label = T, do.return = T, pt.size = 0.5)
```

```{r}
saveRDS(trt.combined.macro, file = file.path(dataPath,"Kulkarni_sce_res40.rds"))
```


#### Differential expression (resolution of 0.40)

Next, we will identify marker genes for 
cluster 0, 1, 2, 3 using the resolution of 0.40

#### Cluster 0

```{r}
clust0.markers <- FindConservedMarkers(trt.combined.macro, ident.1 = 0,
                                       grouping.var = "treatment", 
                                       print.bar = FALSE)
head(clust0.markers)

## remove genes with negative relationship
dat <- clust0.markers[clust0.markers$EnzymeTL_avg_logFC > 0 & 
                      clust0.markers$EnzymeTH_avg_logFC > 0 & 
                      clust0.markers$EnzymeTL_p_val_adj < 0.05 & 
                      clust0.markers$EnzymeTH_p_val_adj < 0.05, ]
dat <- dat[order(rowMeans(dat[, c("EnzymeTL_avg_logFC", "EnzymeTH_avg_logFC")]), 
      decreasing = TRUE),]
dim(dat) # 53 DE genes 
head(dat, 20)

write.csv(dat, file.path(workingPath, 
                         "markergenes/res40/seurat_macrophagesonly_clust0markers.csv"), 
          quote = FALSE, row.names=TRUE)

pdf(file.path(workingPath, 
              "figs/res40/seurat_macrophagesonly_clust0markers.pdf"), 
    width=12, height=12)
FeaturePlot(object = trt.combined.macro, features.plot = row.names(dat)[1:20], 
            min.cutoff = "q9", cols.use = c("lightgrey", "blue"), pt.size = 0.5)
dev.off()

```


#### Cluster 1

```{r}
clust1.markers <- FindConservedMarkers(trt.combined.macro, ident.1 = 1,
                                       grouping.var = "treatment", 
                                       print.bar = FALSE)
head(clust1.markers)

## remove genes with negative relationship
dat <- clust1.markers[clust1.markers$EnzymeTL_avg_logFC > 0 & 
                      clust1.markers$EnzymeTH_avg_logFC > 0 & 
                      clust1.markers$EnzymeTL_p_val_adj < 0.05 & 
                      clust1.markers$EnzymeTH_p_val_adj < 0.05,]
dat <- dat[order(rowMeans(dat[, c("EnzymeTL_avg_logFC", "EnzymeTH_avg_logFC")]), 
      decreasing = TRUE),]
dim(dat) # 141 DE genes 
head(dat, 20)

write.csv(dat, file.path(workingPath, 
                         "markergenes/res40/seurat_macrophagesonly_clust1markers.csv"), 
          quote = FALSE, row.names=TRUE)

pdf(file.path(workingPath, 
              "figs/res40/seurat_macrophagesonly_clust1markers.pdf"), 
    width=12, height=12)
FeaturePlot(object = trt.combined.macro, features.plot = row.names(dat)[1:20], 
            min.cutoff = "q9", cols.use = c("lightgrey", "blue"), pt.size = 0.5)
dev.off()

```



#### Cluster 2

```{r}
clust2.markers <- FindConservedMarkers(trt.combined.macro, ident.1 = 2,
                                       grouping.var = "treatment", 
                                       print.bar = FALSE)
head(clust2.markers)

## remove genes with negative relationship
dat <- clust2.markers[clust2.markers$EnzymeTL_avg_logFC > 0 & 
                      clust2.markers$EnzymeTH_avg_logFC > 0 & 
                      clust2.markers$EnzymeTL_p_val_adj < 0.05 & 
                      clust2.markers$EnzymeTH_p_val_adj < 0.05,]
dat <- dat[order(rowMeans(dat[, c("EnzymeTL_avg_logFC", "EnzymeTH_avg_logFC")]), 
      decreasing = TRUE),]
dim(dat) # 4 DE genes  (really 1)
head(dat, 20)

write.csv(dat, file.path(workingPath, 
                         "markergenes/res40/seurat_macrophagesonly_clust2markers.csv"), 
          quote = FALSE, row.names=TRUE)

pdf(file.path(workingPath, 
              "figs/res40/seurat_macrophagesonly_clust2markers.pdf"), 
    width=5, height=5)
FeaturePlot(object = trt.combined.macro, features.plot = row.names(dat)[1:4], 
            min.cutoff = "q9", cols.use = c("lightgrey", "blue"), pt.size = 0.5)
dev.off()

```



#### Cluster 3

```{r}
clust3.markers <- FindConservedMarkers(trt.combined.macro, ident.1 = 3,
                                       grouping.var = "treatment", 
                                       print.bar = FALSE)
head(clust3.markers)

## remove genes with negative relationship
dat <- clust3.markers[clust3.markers$EnzymeTL_avg_logFC > 0 & 
                      clust3.markers$EnzymeTH_avg_logFC > 0 & 
                      clust3.markers$EnzymeTL_p_val_adj < 0.05 & 
                      clust3.markers$EnzymeTH_p_val_adj < 0.05,]
dat <- dat[order(rowMeans(dat[, c("EnzymeTL_avg_logFC", "EnzymeTH_avg_logFC")]), 
      decreasing = TRUE),]
dim(dat) # XXXX DE genes 
head(dat, 20)

write.csv(dat, file.path(workingPath, 
                         "markergenes/res40/seurat_macrophagesonly_clust3markers.csv"), 
          quote = FALSE, row.names=TRUE)

pdf(file.path(workingPath, 
              "figs/res40/seurat_macrophagesonly_clust3markers.pdf"), 
    width=12, height=12)
FeaturePlot(object = trt.combined.macro, features.plot = row.names(dat)[1:20], 
            min.cutoff = "q9", cols.use = c("lightgrey", "blue"), pt.size = 0.5)
dev.off()

```

