---
title: "Figures for 2018 SPARC Proposal"
author: "Stephanie Hicks"
output: 
    html_document:
        toc: true
        toc_float: true
        highlight: tango
        number_sections: true
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval=FALSE)
```

# Preprocessing and Normalization

Load packages
```{r loadpackages, message=FALSE, warning=FALSE, eval=TRUE}
dataPath <- "/users/shicks1/data/kulkarni"
workingPath <- "/users/shicks1/projects/kulkarni"
```

## Data


Load data and create a `SingleCellExperiment` object. 
Remove genes not expressed in any cell. Extract `mgi_symbol` 
gene name`. 
```{r sceobject, eval=FALSE}
library(monocle)
library(scater)
cds <- readRDS(file.path(dataPath, "Kulkarni_cds.rds")) # CellDataSet class in monocle

# Convert to a SingleCellExperiment
gene_df <- DataFrame(Gene = rownames(exprs(cds)))
rownames(gene_df) <- gene_df$Gene
sce <- SingleCellExperiment(assays = list(counts = exprs(cds)), 
                            colData = pData(cds), rowData = gene_df)

# Remove genes not expressed in any cell
keep_feature <- rowSums(counts(sce) > 0) > 0
table(keep_feature)
sce <- sce[keep_feature, ]

# calcualte QC metrics
sce <- calculateQCMetrics(sce)
head(colData(sce))

# extract mgi gene symbol name
sce <- scater::getBMFeatureAnnos(sce,
    filters = "ensembl_gene_id",
    attributes = c("ensembl_gene_id",
        "mgi_symbol", "chromosome_name",
        "start_position", "end_position"),
    feature_symbol = "mgi_symbol",
    feature_id = "ensembl_gene_id",
    biomart = "ENSEMBL_MART_ENSEMBL",
    dataset = "mmusculus_gene_ensembl",
    host = "www.ensembl.org")

# define feature names in `feature_symbol` column
rownames(sce) <- rowData(sce)$feature_symbol
saveRDS(sce, file = file.path(dataPath,"Kulkarni_sce.rds"))
```

Convert to Seurat
```{r}
library(Seurat)
sce <- readRDS(file.path(dataPath,"Kulkarni_sce.rds")) 
sce <- sce[!duplicated(row.names(sce)), ]

# Set up treatment (TH) object
trtTH <- CreateSeuratObject(raw.data = SummarizedExperiment::assay(sce[,
                            SummarizedExperiment::colData(sce)$treatment == "TH"], "counts"), 
                            project = "TRT_TH", min.cells = 5)
trtTH@meta.data <- data.frame(trtTH@meta.data, 
                              as.data.frame(SummarizedExperiment::colData(sce[,
                                        SummarizedExperiment::colData(sce)$treatment == "TH"])))
trtTH <- FilterCells(trtTH, subset.names = "nGene", low.thresholds = 500, high.thresholds = Inf)
trtTH <- NormalizeData(trtTH)
trtTH <- ScaleData(trtTH, display.progress = F)

# Set up treatment (TL) object
trtTL <- CreateSeuratObject(raw.data = SummarizedExperiment::assay(sce[,
                                        SummarizedExperiment::colData(sce)$treatment == "TL"], "counts"), 
                            project = "TRT_TL", min.cells = 5)
trtTL@meta.data <- data.frame(trtTL@meta.data, 
                              as.data.frame(SummarizedExperiment::colData(sce[,
                                        SummarizedExperiment::colData(sce)$treatment == "TL"])))
trtTL <- FilterCells(trtTL, subset.names = "nGene", low.thresholds = 500, high.thresholds = Inf)
trtTL <- NormalizeData(trtTL)
trtTL <- ScaleData(trtTL, display.progress = F)
```

Gene selection for input to CCA
```{r}
trtTH <- FindVariableGenes(trtTH, do.plot = F)
trtTL <- FindVariableGenes(trtTL, do.plot = F)
g.1 <- head(rownames(trtTH@hvg.info), 1000)
g.2 <- head(rownames(trtTL@hvg.info), 1000)
genes.use <- unique(c(g.1, g.2))
genes.use <- intersect(genes.use, rownames(trtTH@scale.data))
genes.use <- intersect(genes.use, rownames(trtTL@scale.data))
```

Perform CCA. Identifies common sources of variation between two data sets. 
the `RunCCA()` function will also combine two objects into a single object 
and store the canonical correlation vectors (the vectors that project each 
dataset into the maximally correlated subspaces). We also store the 
original dataset identity as a column in `object@meta.data`. 
```{r}
trt.combined <- RunCCA(trtTH, trtTL, genes.use = genes.use, num.cc = 30)

trt.combined@meta.data$treatment <- 
  factor(trt.combined@meta.data$treatment, 
         levels=c("TH", "TL"), labels=c("Enzyme 1", "Enzyme 2"))
```


Align the CCA subspaces
```{r}
trt.combined <- AlignSubspace(trt.combined, reduction.type = "cca", grouping.var = "treatment", 
    dims.align = 1:25)
```

Visualize T-SNE
```{r}
trt.combined <- RunTSNE(trt.combined, reduction.use = "cca.aligned", dims.use = 1:25, 
    do.fast = T)
```

Resolution = 0.25
```{r}
trt.combined <- FindClusters(trt.combined, reduction.type = "cca.aligned", 
    resolution = 0.25, dims.use = 1:25, force.recalc=TRUE)
```

Create t-SNE figure with all small intestine cell types
```{r Figure}
new.ident <- c("Macrophages", "Mesothelial Cells", "Neuroglia",
         "Perivascular\nEndothelial Cells","B Lymphocytes", 
         "Adipocytes", "Pericytes", "T and NK\n    Lymphocytes", 
         "             Enterocytes", "Contractile Smooth\nMuscle Cells", 
         "Erythrocytes")
for (i in 0:10) {
    trt.combined <- RenameIdent(object = trt.combined, old.ident.name = i, 
        new.ident.name = new.ident[i + 1])
}
```

For 2019-03-17-Discovery figure
```{r}
p1 <- TSNEPlot(trt.combined, do.label = T, do.return = T, pt.size = 0.5, 
               no.legend = TRUE, label.size= 4,
               plot.title = "scRNA-Seq Analysis of 4477 cells from LM-MP") + 
   xlab("t-SNE Dimension 1") + ylab("t-SNE Dimension 2") + 
  theme(plot.title = element_text(size = 16), 
        axis.text=element_text(size=12),
        axis.title=element_text(size=16))
p2 <- TSNEPlot(trt.combined, do.return = T, pt.size = 0.5, group.by = "treatment", 
               plot.title = "Two dissociation enzymes") + 
        xlab("t-SNE Dimension 1") + ylab("t-SNE Dimension 2") +
        theme(plot.title = element_text(size = 16), legend.position=c(0.9, 0.9),
              # legend.text = element_text(size = 8),
          axis.text=element_text(size=12), axis.title=element_text(size=16))
p3 <- FeaturePlot(object = trt.combined, features.plot = c("Ccr2"), do.return=TRUE,
            min.cutoff = "q9", cols.use = c("lightgrey", "blue"), pt.size = 0.5) 
p3 <- p3$Ccr2 + xlab("t-SNE Dimension 1") + ylab("t-SNE Dimension 2") + 
  theme(axis.text=element_text(size=12), axis.title=element_text(size=16), 
        plot.title = element_text(size=16, face="plain")) + 
  ggtitle("Macrophage Population 1 (Gene: Ccr2)")

p4 <- FeaturePlot(object = trt.combined, features.plot = c("Cd163"), do.return=TRUE,
            min.cutoff = "q9", cols.use = c("lightgrey", "blue"), pt.size = 0.5) 
p4 <- p4$Cd163 + xlab("t-SNE Dimension 1") + ylab("t-SNE Dimension 2") + 
  theme(axis.text=element_text(size=12), axis.title=element_text(size=16), 
        plot.title = element_text(size=16, face="plain")) + 
  ggtitle("Macrophage Population 2 (Gene: Cd163)")

p5 <- FeaturePlot(object = trt.combined, features.plot = c("Irak1"), do.return=TRUE,
            min.cutoff = "q9", cols.use = c("lightgrey", "blue"), pt.size = 0.5) 
p5 <- p5$Irak1 + xlab("t-SNE Dimension 1") + ylab("t-SNE Dimension 2") + 
  theme(axis.text=element_text(size=12), axis.title=element_text(size=16), 
        plot.title = element_text(size=16, face="plain")) + 
  ggtitle("Gene: Irak1")


fig <- ggdraw() + 
    draw_plot(p1, x = 0, y = .5, width = .50, height = .5) +
    draw_plot(p3, x = .50, y = .5, width = .5, height = .5) +
    draw_plot(p5, x = 0, y = 0, width = .5, height = .5) +
    draw_plot(p4, x = .5, y = 0, width = .5, height = .5) +
    draw_plot_label(label = LETTERS[1:4], size = 25,
                    x = c(0, 0.5, 0, 0.5), y = c(1, 1, 0.5, 0.5))
pdf(file.path(workingPath, "2019-03-17-Discovery-Figure-celltype.pdf"), width=12, height=10)
print(fig)
dev.off()
```


2019-03-29 Discovery figure 

Looks like the macrophages are labeled as the `0` 
population. Next, filter for only macrophages and 
create a separate Seurat object
```{r}
trt.combined.macro <- SubsetData(trt.combined, 
           cells.use = trt.combined@cell.names[trt.combined@meta.data$res.0.25 == 0])
```

Visualize T-SNE
```{r}
trt.combined.macro <- RunTSNE(trt.combined.macro, reduction.use = "cca.aligned", dims.use = 1:25, 
    do.fast = T)
TSNEPlot(trt.combined.macro, do.return = T, pt.size = 0.5, group.by = "treatment")
TSNEPlot(trt.combined.macro, do.return = T, pt.size = 0.5, group.by = "cell_type")
TSNEPlot(trt.combined.macro, do.label = T, do.return = T, pt.size = 0.5)
```


```{r}
p1 <- FeaturePlot(object = trt.combined.macro, features.plot = c("Ccr2"), do.return=TRUE,
            min.cutoff = "q9", cols.use = c("lightgrey", "blue"), pt.size = 0.5) 
p1 <- p1$Ccr2 + xlab("t-SNE Dimension 1") + ylab("t-SNE Dimension 2") + 
  theme(axis.text=element_text(size=12), axis.title=element_text(size=16), 
        plot.title = element_text(size=16, face="plain")) + 
  ggtitle("Macrophage Population 1\n(Gene: Ccr2)")

p2 <- FeaturePlot(object = trt.combined.macro, features.plot = c("Cd163"), do.return=TRUE,
            min.cutoff = "q9", cols.use = c("lightgrey", "blue"), pt.size = 0.5) 
p2 <- p2$Cd163 + xlab("t-SNE Dimension 1") + ylab("t-SNE Dimension 2") + 
  theme(axis.text=element_text(size=12), axis.title=element_text(size=16), 
        plot.title = element_text(size=16, face="plain")) + 
  ggtitle("Macrophage Population 2\n(Gene: Cd163)")

library(cowplot)
fig <- plot_grid(p1, p2, labels = LETTERS[1:2], label_size = 20)
pdf(file.path(workingPath, "2019-03-29-Discovery-Figure-celltype.pdf"), width=8, height=4)
print(fig)
dev.off()

```
